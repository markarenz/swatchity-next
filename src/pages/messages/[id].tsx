import React, { useState } from 'react';
import { Message } from '@prisma/client';
import type { NextPage, GetServerSideProps } from 'next';
import { getServerSession } from 'next-auth/next';
import { authOptions } from '@/pages/api/auth/[...nextauth]';
import { useUserContext } from '@/context/UserContext';
import Layout from '@/components/pageComponents/Layout';
import { useIntl, FormattedMessage } from 'react-intl';
import UserProfileBlock from '@/components/common/UserProfileBlock';
import MessagePost from '@/components/common/MessagePost';
import { useRouter } from 'next/router';
import MessageSkeleton from '@/components/common/MessageSkeleton';

import {
  serializeMessageDates,
  serializeUserMeta,
  getReplies,
  createReply,
  getMessages,
} from '@/utils/apiFunctions';
import { getMessagesDB } from '@/utils/dbFunctions';
import { PageMeta, UserProfile } from '@/types';
import ButtonIcon from '@/components/common/ButtonIcon';
import IconRefresh from '@/components/icons/IconRefresh';
import { repliesPerPage } from '@/constants';
import ReplySkeleton from '@/components/common/ReplySkeleton';

type Props = {
  initialMessages: Message[];
  otherUser: UserProfile;
};
const MessageThreadPage: NextPage<Props> = ({ initialMessages, otherUser }) => {
  const [isLoading, setIsLoading] = useState(false);
  const [isCreating, setIsCreating] = useState(false);
  const [canLoadMore, setCanLoadMore] = useState(initialMessages.length > 0);
  const [messages, setMessages] = useState(initialMessages);
  const [isReplyOpen, setIsReplyOpen] = useState(false);
  const { userMeta, checkUserMeta } = useUserContext();
  const router = useRouter();
  const { formatMessage } = useIntl();
  const pageMeta: PageMeta = {
    title: 'Message Thread',
    metadesc: 'Have an in-depth conversation the Swatchity way!',
  };
  const isLoggedIn = !!userMeta?.name;
  const refreshMessages = () => {
    setCanLoadMore(true);
    setMessages([]);
    loadMessages(0);
  };
  const loadMoreMessages = () => {
    loadMessages(messages.length);
  };
  const loadMessages = async (sk: number) => {
    setIsLoading(true);
    const newMessagesData = await getMessages(`${otherUser?.id}`, sk);
    const { messages: newMessages } = newMessagesData;
    if (sk > 0) {
      setMessages([...messages, ...newMessages]);
    } else {
      setMessages(newMessages);
    }
    if (newMessages.length === 0) {
      setCanLoadMore(false);
    }
    setIsLoading(false);
  };

  const headerButtons = (
    <React.Fragment>
      <ButtonIcon
        onClick={refreshMessages}
        label={formatMessage({ id: 'feed__refresh' })}
        testID="btn-refresh"
      >
        <div key="btn-refresh" className="hover-zoom h-3 w-3 py-0-5 px-0-5">
          <IconRefresh color="gray-6" colorDark="gray-2" />
        </div>
      </ButtonIcon>
    </React.Fragment>
  );
  const selfName = `${userMeta?.name}`;
  const otherName = `${otherUser?.name}`;
  if (!isLoggedIn) {
    return (
      <Layout pageMeta={pageMeta} headerButtons={headerButtons}>
        <div id="message-thread">
          <div className="contained pt-1 pb-4">
            <h2>
              <FormattedMessage id="messages__not_logged_in" />
            </h2>
          </div>
        </div>
      </Layout>
    );
  }
  return (
    <Layout pageMeta={pageMeta} headerButtons={headerButtons}>
      <div id="message-thread">
        <div className="contained pt-1 pb-4">
          <UserProfileBlock userProfile={otherUser} mode="messages" setMessages={setMessages} />
          <div id="messages-feed">
            {isCreating && <ReplySkeleton idx={0} />}
            {messages.map((m) => (
              <MessagePost
                key={m.id}
                message={m}
                isSelf={m.fromUserID === userMeta?.id}
                name={m.fromUserID === userMeta?.id ? selfName : otherName}
              />
            ))}

            {isLoading &&
              [...Array(repliesPerPage)].map((_e, idx) => (
                <MessageSkeleton key={`skeleton-message-${idx}`} isSelf={idx % 2 === 0} idx={idx} />
              ))}
            {!isLoading && messages.length === 0 && (
              <h2 className="pl-5">
                <FormattedMessage id="messages__thread__no_messages" />
              </h2>
            )}
            {canLoadMore && (
              <div className="w-full py-1">
                <button
                  onClick={loadMoreMessages}
                  className="btn margin-center"
                  data-testid="feed-load-more"
                >
                  <FormattedMessage id="feed__load_more" />
                </button>
              </div>
            )}
          </div>
        </div>
      </div>
    </Layout>
  );
};

export const getServerSideProps: GetServerSideProps = async (context) => {
  const session = await getServerSession(context.req, context.res, authOptions);
  const { id } = context.query;
  const messagesData = await getMessagesDB(session, `${id}`, 0);
  const { messages, otherUserProfile } = messagesData;
  const initialMessages = serializeMessageDates(messages);
  return {
    props: {
      initialMessages,
      otherUser: otherUserProfile,
    },
  };
};

export default MessageThreadPage;
