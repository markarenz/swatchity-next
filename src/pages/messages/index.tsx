import React, { useState, useEffect } from 'react';
import Link from 'next/link';
import type { NextPage, GetServerSideProps } from 'next';
import { getServerSession } from 'next-auth/next';
import { authOptions } from '@/pages/api/auth/[...nextauth]';
import { useUserContext } from '@/context/UserContext';
import { PageMeta } from '@/types';
import { serializeMessageThreads } from '@/utils/apiFunctions';
import { getMessageThreads, getSwatchesDB } from '@/utils/dbFunctions';
import Layout from '@/components/pageComponents/Layout';
import { MessageThread } from '@prisma/client';
import { FormattedMessage } from 'react-intl';
import IconLink from '@/components/icons/IconLink';
import MessageThreadListItem from '@/components/common/MessageThreadListItem';

type Props = {
  threads: MessageThread[];
};
const MessageThreads: NextPage<Props> = ({ threads }) => {
  const [lastVisit, setLastVisit] = useState<string | null>(null);
  const [pageInit, setPageInit] = useState(false);
  const { userMeta } = useUserContext();
  const isLoggedIn = !!userMeta?.name;
  const pageMeta: PageMeta = {
    title: 'Messages',
    metadesc:
      'Connect with your fellow Swatchers! To start a message thread, just click on Message in the profile of a friend.',
  };

  useEffect(() => {
    if (!pageInit) {
      setPageInit(true);
      if (typeof window !== 'undefined' && lastVisit === null) {
        const newLastVisit = localStorage.getItem('lastMessagesView') || '';
        setLastVisit(newLastVisit);
      }
    }
    return () => {
      if (pageInit) {
        if (typeof window !== 'undefined') {
          localStorage.setItem('lastMessagesView', new Date().toISOString());
        }
      }
    };
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [pageInit]);

  return (
    <Layout pageMeta={pageMeta}>
      <div className="contained pt-1 pb-4" id="message-threads">
        <h1 className="text-3">{pageMeta.title}</h1>
        <div className="mb-1 italic">
          <p>{pageMeta.metadesc}</p>
        </div>
        {!isLoggedIn && (
          <h2>
            <FormattedMessage id="messages__not_logged_in" />
          </h2>
        )}
        {isLoggedIn && (
          <div id="threads-feed">
            {threads.map((thread) => (
              <MessageThreadListItem
                key={thread.id}
                thread={thread}
                userID={userMeta?.id}
                lastVisitStr={`${lastVisit}`}
              />
            ))}
          </div>
        )}
      </div>
    </Layout>
  );
};

export const getServerSideProps: GetServerSideProps = async (context) => {
  const session = await getServerSession(context.req, context.res, authOptions);
  const threadsData = await getMessageThreads(session);
  const { threads: threadsRaw } = threadsData;
  const threads = serializeMessageThreads(threadsRaw);
  return {
    props: {
      threads,
    },
  };
};

export default MessageThreads;
