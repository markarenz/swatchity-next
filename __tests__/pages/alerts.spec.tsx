import { render, screen, act, fireEvent } from '@testing-library/react';
import { GetServerSidePropsContext } from 'next';
import AlertsPage, { getServerSideProps } from '@/pages/alerts';
import { createRequest, createResponse } from 'node-mocks-http';
import { IntlProvider } from 'react-intl';
import messages from '@/locale/en-US.json';
import mockSwatch from '../__fixtures__/mockSwatch';
import mockUserData from '../__fixtures__/mockUserMeta';
import mockAlert from '../__fixtures__/mockAlert';

jest.useFakeTimers().setSystemTime(new Date('2023-03-04T22:22:51.201+00:00'));

const initialLikes = ['abcd1234'];
const initialSwatches = [mockSwatch];
jest.mock('@/components/common/SwatchFeed', () => () => <div>Swatch Feed</div>);

jest.mock('next', () => ({
  ...jest.requireActual('next'),
  GetServerSideProps: () => () => {},
}));

jest.mock('next-auth/next', () => ({
  getServerSession: jest.fn(() => ({
    expires: new Date(Date.now() + 2 * 86400).toISOString(),
    user: { name: 'test', email: 'test@domain.com' },
  })),
}));

jest.mock('@/context/UserContext', () => ({
  useUserContext: jest.fn(() => ({
    userMeta: mockUserData,
  })),
}));

jest.mock('@/utils/dbFunctions', () => ({
  getAlertsDB: jest.fn(() => ({
    alerts: [mockAlert],
  })),
}));
jest.mock('@/utils/apiFunctions', () => ({
  ...jest.requireActual('@/utils/apiFunctions'),
  getAlerts: jest
    .fn()
    .mockImplementation(() => ({
      alerts: [mockAlert],
    }))
    .mockImplementationOnce(() => ({
      alerts: [],
    })),
  // getAlerts2: jest.fn(() => ({
  //   alerts: [mockAlert],
  // })),
}));

jest.mock('next-auth', () => ({
  __esModule: true,
  default: () =>
    Promise.resolve({
      NextAuth: jest.fn(),
    }),
}));

jest.mock('next-auth/react', () => ({
  useSession: jest.fn().mockReturnValue({
    session: {
      email: 'email@test.com',
    },
  }),
}));

jest.mock('next/router', () => ({
  useRouter: jest.fn(() => ({
    pathname: '/',
  })),
}));

describe('Alert Page Page', () => {
  it('renders page', () => {
    act(() => {
      render(
        <IntlProvider messages={messages} locale="en" defaultLocale="en">
          <AlertsPage initialAlerts={[mockAlert]} />
        </IntlProvider>,
      );
    });
    expect(screen).toMatchSnapshot();
  });

  it('renders page - has unread alerts', () => {
    Storage.prototype.getItem = jest.fn().mockReturnValue('TEST');
    act(() => {
      render(
        <IntlProvider messages={messages} locale="en" defaultLocale="en">
          <AlertsPage initialAlerts={[mockAlert]} />
        </IntlProvider>,
      );
    });
    expect(screen).toMatchSnapshot();
  });

  it('handles load more', async () => {
    act(() => {
      render(
        <IntlProvider messages={messages} locale="en" defaultLocale="en">
          <AlertsPage initialAlerts={[mockAlert]} />
        </IntlProvider>,
      );
    });
    const btnMore = await screen.findByTestId('feed-load-more');
    act(() => {
      fireEvent(
        btnMore,
        new MouseEvent('click', {
          bubbles: true,
          cancelable: true,
        }),
      );
    });
    expect(screen).toMatchSnapshot();
  });

  it('handles load more - no results', async () => {
    act(() => {
      render(
        <IntlProvider messages={messages} locale="en" defaultLocale="en">
          <AlertsPage initialAlerts={[mockAlert]} />
        </IntlProvider>,
      );
    });
    const btnMore = await screen.findByTestId('feed-load-more');
    act(() => {
      fireEvent(
        btnMore,
        new MouseEvent('click', {
          bubbles: true,
          cancelable: true,
        }),
      );
    });
    expect(screen).toMatchSnapshot();
  });

  it('handles refresh', async () => {
    act(() => {
      render(
        <IntlProvider messages={messages} locale="en" defaultLocale="en">
          <AlertsPage initialAlerts={[mockAlert]} />
        </IntlProvider>,
      );
    });
    const btnRefresh = await screen.findByTestId('btn-refresh');
    act(() => {
      fireEvent(
        btnRefresh,
        new MouseEvent('click', {
          bubbles: true,
          cancelable: true,
        }),
      );
    });
    expect(screen).toMatchSnapshot();
  });

  it('GetServerSideProps', async () => {
    const req = createRequest({
      method: 'GET',
    });
    const res = createResponse();
    const context = {
      req,
      res,
    };
    const result = await getServerSideProps(context as GetServerSidePropsContext);
    const expectedResult = JSON.stringify({
      props: {
        initialAlerts: [mockAlert],
      },
    });
    expect(JSON.stringify(result)).toEqual(expectedResult);
  });
});
