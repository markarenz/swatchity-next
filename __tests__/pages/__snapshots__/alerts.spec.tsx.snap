// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Alert Page Page handles load more - no results 1`] = `
{
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "logTestingPlaygroundURL": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
}
`;

exports[`Alert Page Page handles load more 1`] = `
{
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "logTestingPlaygroundURL": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
}
`;

exports[`Alert Page Page handles refresh 1`] = `
{
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "logTestingPlaygroundURL": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
}
`;

exports[`Alert Page Page renders page - has unread alerts 1`] = `
{
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "logTestingPlaygroundURL": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
}
`;

exports[`Alert Page Page renders page 1`] = `
{
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "logTestingPlaygroundURL": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
}
`;
